### ISSUE TITLE
Reorganize the code base directories

### ISSUE THREAD

**Issue by sergii-iakovenko** [telstra/open-kilda#1137]
Currently, the code base is scattered among several modules which have unclear dependency management and ineffective grouping. Reorganization can improve developer experience, provide better code cohesion and dependency management.

Goals / Requirements:
- Deployable system components must be isolated from others. E.g. Northbound is a deployable service, which must not be a part of any component module.
- Module contains classes which correspond to the module scope. E.g. Northbound DTO classes designed to be used by Northbound only must NOT appear in a topology module.
- Prefer hierarchical over flat layout.
- A developer should require to locate only the root parent pom in order to open the code base in IDE.

WBS:
- [x] Regroup and move system components to `src-java` and `src-python` folders
- [x] Split WFM into multiple topology-focused modules
- [x] Migrate to Gradle incremental build with single root project for Java-based modules. 
- [x] Collect Dockerfiles in `docker` folder
- [x] Update Dockerfiles and docker-compose to support incremental build of docker images
- [x] Update Makefile to provide incremental build options
- [x] Update Travis jobs
- [x] Update Jenkins jobs
- [x] Move GUI to `src-gui`

Proposed structure:
```
  │
  ├── confd                       # Directory with Confd templates
  │    └...
  │
  ├── docker                      # Directory with Dockerfile(s)
  │    ├── base
  │    ├── floodlight-modules     # Dockerfile for Floodlight-Modules component
  │    ├── kafka                  # Dockerfile for Kafka image (applicable to all-in-one environemnts)
  │    ├── neo4j                  # Dockerfile for Neo4j image (applicable to all-in-one environemnts)
  │    ├── network-topology       # Dockerfile for Network Topology submitter
  │    ├── flow-topology          # Dockerfile for Flow Topology submitter
  │    └...
  │
  ├── docs
  │    ├── design                 # Project and feature design documentation
  │    ├── guidelines             # OpenKilda developer's guidelines and conventions
  │    └...
  │  
  ├── src-gui                     # OpenKilda GUI source code
  │    ├── README.md
  │    └...
  │  
  ├── src-java                    # Source code of Java-based components
  │    ├── build.gradle
  │    ├── settings.gradle
  │    ├── README.md
  │    │
  │    ├── checkstyle             # Project-wide checkstyle plugin configuration
  │    ├── kilda-configuration    
  │    │    ├── build.gradle
  │    │    ├── README.md
  │    │    └── src
  │    │         ├── checkstyle   # Module specific checkstyle plugin configuration
  │    │         ├── main         # Module production source code
  │    │         └── test         # Module unit tests
  │    │
  │    ├── northbound-service
  │    │    ├── README.md
  │    │    ├── northbound-api
  │    │    │    ├── build.gradle
  │    │    │    ├── README.md
  │    │    │    └── src
  │    │    │         ├── checkstyle   # Module specific checkstyle plugin configuration
  │    │    │         ├── main         # Module production source code
  │    │    │         └── test         # Module unit tests
  │    │    └── northbound
  │    │         ├── build.gradle
  │    │         ├── README.md
  │    │         └── src
  │    │              ├── checkstyle   # Module specific checkstyle plugin configuration
  │    │              ├── main         # Module production source code
  │    │              └── test         # Module unit tests
  │    │
  │    ├── base-topology               # Shared components / classes for Storm topologies 
  │    │    ├── base-messaging
  │    │    │    └...
  │    │    └── base-storm-topology
  │    │         └...
  │    │
  │    ├── network-topology            # Network topology source code
  │    │    ├── network-messaging
  │    │    │    └...
  │    │    └── network-storm-topology
  │    │         └...
  │    │
  │    ├── testing                     # Functional, ATDD and other tests
  │    │    ├── test-library
  │    │    │    └...
  │    │    ├── functional-tests
  │    │    │    └...
  │    │    └── performance-tests
  │    │         └...
  │    └...
  │  
  ├── src-python                  # Source code of Python-based components.
  │    ├── lab-service            # Lab-service's source code.
  │    │    ├── README.md
  │    │    └...
  │    │
  │    └── lock-keeper            # Lock Keeper's source code
  │         └...
  │  
  ├── tools                       # Various tools and utilities used in the development and support process
  │    └...
  │  
  ├── Makefile                    # Script for local builds
  ├── docker-compose.yml
  ├── Jenkinsfile                 # Scripts used for Continuous Integration
  ├── .travis.yml                 # Scripts used for Continuous Integration
  │  
  ├── CHANGELOG.md                # A log of made changes and release notes
  ├── LICENSE.txt
  └── README.md
```

**Comment by abochkarev** [comment]
@sergii-iakovenko In general, looks good, but I'd like to clarify the following:
- Is it correct that all module dependents tests (i.e: unit tests) will be located in the **testing** module? If I'm not right, please clarify.
- Is it possible to apply a common naming pattern for modules with API? I mean: northbound-api & northbound-core, floodlight-kilda-api & floodlight-kilda-core etc?
- Where is planned to keep checkstyle configs, doccumentation (if needed)?

**Comment by abochkarev** [comment]
@sergii-iakovenko Thank you for the detailed clarification. No more questions/suggestions from my side.

**Comment by timofei-durakov** [comment]
overall looks good, one question from my side:
there are common/messaging and storm-topologies/messaging packages. So is it really possible to split them up, or could their objects overlap?

**Comment by surabujin** [comment]
Why project wide documentation are stored into src-java folder?
Can we start migrating to this schema?

**Comment by sergii-iakovenko** [comment]
@neeraj890kumar Please, link the PR with "Move GUI to src-gui" changes to this ticket.

**Comment by sergii-iakovenko** [comment]
Resolved by PRs: #3095, #3202, #3327.
